src/requestforattestation/RequestForAttestation.ts

```diff
@@ -107,22 +113,55 @@ export default class RequestForAttestation implements IRequestForAttestation {
    */
   public constructor(
     claim: IClaim,
-    legitimations: AttestedClaim[],
-    identity: Identity,
+    legitimations: IAttestedClaim[],
+    identity?: Identity,
+    claimOwner?: NonceHash,
+    claimerSignature?: string,
+    claimHashTree?: object,
+    ctypeHash?: NonceHash,
+    rootHash?: Hash,
     delegationId?: IDelegationBaseNode['id']
   ) {
-    if (claim.owner !== identity.address) {
-      throw Error('Claim owner is not identity')
-    }
-    this.claim = claim
-    this.claimOwner = generateHash(this.claim.owner)
-    this.ctypeHash = generateHash(this.claim.cType)
-    this.legitimations = legitimations
-    this.delegationId = delegationId
+    if (identity !== undefined) {
+      if (claim.owner !== identity.address) {
+        throw Error('Claim owner is not identity')
+      }
+      this.claim = claim
+      this.claimOwner = generateHash(this.claim.owner)
+      this.ctypeHash = generateHash(this.claim.cTypeHash)
+      this.legitimations = legitimations.map((legitimation: IAttestedClaim) =>
+        AttestedClaim.fromObject(JSON.parse(
+          JSON.stringify(legitimation)
+        ) as IAttestedClaim)
+      )
+      this.delegationId = delegationId
 
-    this.claimHashTree = generateHashTree(claim.contents)
-    this.hash = this.calculateRootHash()
-    this.claimerSignature = this.sign(identity)
+      this.claimHashTree = generateHashTree(claim.contents)
+      this.rootHash = this.calculateRootHash()
+      this.claimerSignature = this.sign(identity)
```

Comment:
If we have a separate method for building this RequestForAttestation from a claim, identity and legitimations, this hash and hashtree generation could live there. In the constructor, we would just need to verify the signatures

Reply:
separate into from...() helper functions